src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs

```diff
@@ -93,28 +104,73 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
             return description;
         }
 
+        internal static async Task<IsMasterResult> GetIsMasterResultAsync(
+            IConnection connection,
+            CommandWireProtocol<BsonDocument> isMasterProtocol,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                return new IsMasterResult(await isMasterProtocol
+                    .ExecuteAsync(connection, cancellationToken).ConfigureAwait(false));
+            }
+            catch (MongoCommandException ex)
+            {
+                // If the isMaster command fails with error code 11 (UserNotFound), drivers must consider authentication
+                // to have failed.In such a case, drivers MUST raise an error that is equivalent to what they would have
+                // raised if the authentication mechanism were specified and the server responded the same way.
+                if (ex.Code == 11)
+                {
+                    throw new MongoAuthenticationException(connection.ConnectionId, "User not found.", ex);
```

Comment:
This exception currently gets "lost in the abyss": we get a timeout exception as previously discussed.